// _stores/story.js
var Reflux = require('reflux');
var Actions = require('_actions');
var Promise = require('bluebird');
var t = require('_i18n');
var ffService = require('_services/fanfiction');

var acceptedDomains = require('_config').acceptedDomains;

var storyStore = Reflux.createStore({
    init: function() {
        this.checking = false;
        this.listenTo(Actions.checkStory, this.checkStory);
    },
    checkStory: function(_url) {
        this._setStatus(false, 'fetching', true);
        ffService.check(_url).bind(this).then(function(story) {
            this._setStatus(false, 'success', false);
        }).catch(function(e) {
            this._setStatus(e.getMessage());
        });
    },
    _setStatus: function(_error, _trigger, _checking) {
        this.lastIsError = _error || false;
        this.checking = _checking || false;
        var trigger = _trigger || 'error';
        this.trigger(trigger);
    },
    getStatus: function() {
        return {
            fetching: this.checking,
            error: this.lastIsError
        };
    }
});

module.exports = storyStore;
