// _stores/story.js
var Reflux = require('reflux');
var Actions = require('_actions');
var ffService = require('_services/fanfiction');

//@todo factorize all the internal state ...
var storyStore = Reflux.createStore({
    init: function() {
        this.lastInput = null;
        this.fetching = false;
        this.story = {};
        this.lastIsError = false;
        this.listenTo(Actions.checkStory, this.checkStory);
    },
    checkStory: function(_url) {
        this.lastInput = _url;
        this.lastIsError = false;
        this.fetching = true;
        this.trigger('fetching');
        ffService.check(_url).bind(this).then(function(story) {
            console.log(story);
            this.story = story;
            this.fetching = false;
            this.lastIsError = false;
            this.trigger('success');
        }).catch(function(e) {
            this.story = {};
            this.fetching = false;
            this.lastIsError = e.message;
            this.trigger('error');
        });
    },
    getStatus: function() {
        return {
            fetching: this.fetching,
            error: this.lastIsError,
            input: this.lastInput,
            story: this.story
        };
    }
});

module.exports = storyStore;
