// _components/Story.jsx
var React = require('react/addons');
var t = require('_i18n');
var moment = t.moment;
var _ = require('underscore');

var dateFormat = 'dddd D MMMM YYYY';

var Story = React.createClass({
    mixins: [React.addons.PureRenderMixin],
    propTypes: {
        fetching: React.PropTypes.bool,
        story: React.PropTypes.object,
        error: React.PropTypes.string
    },
    getDefaultProps: function() {
        return {
            story: {},
            fetching: false,
            error: ''
        };
    },
    _conditionalFetching: function() {
        if (this.props.fetching) {
            return <div className="loading">{t.gettext('Loading ...')}</div>
        }
    },
    _conditionalError: function() {
        if (this.props.error) {
            return this.props.error;
        }
    },
    _conditionalDetails: function() {
        if (_.isEmpty(this.props.story)) return null;

        return (
        <div className="story-details">
            <div>{t.gettext('Story : ')}{this.props.story.storyName}</div>
            <div>{t.gettext('Author : ')}{this.props.story.authorName}</div>
            <div>{t.ngettext('Chapter : %d', 'Chapters : %d', this.props.story.chaptersCount)}</div>
            <div>{t.gettext('Published : ')}{moment(this.props.story.published).format(dateFormat)}</div>
            <div>{t.gettext('Last updated : ')}{moment(this.props.story.lastUpdated).format(dateFormat)}</div>
            <div>{t.gettext('Status : ')}{this.props.story.storyStatus}</div>
            <div>{t.gettext('Summary : ')}{this.props.story.summary}</div>
            <div>{t.gettext('Genre : ')}{this.props.story.genre}</div>
            <div>{t.gettext('Category : ')}{this.props.story.category}</div>
        </div>
        );
    },
    render: function() {
        return (
        <div className="story">
            <div>
                {this._conditionalFetching()}
                {this._conditionalError()}
            </div>
            {this._conditionalDetails()}
        </div>
        );
    }
});

module.exports = Story;
