// _services/adapters/www.fanfiction.net.adapter.js
var Promise = require('bluebird');

function process(source) {
    var websiteName = 'www.FanFiction.net';
    var storyName = '';
    var authorName = '';
    var chaptersCount = 0;
    var lastUpdated = '';
    var summary = '';
    var storyStatus = 'Unknown';
    var category = '';
    var genre = '';
    var rating = '';
    var storyLink = '';
    var authorLink = '';
    var totalWordCount = 0;
    var chapterNames = [];
    var chapterLinks = [];

    return {
        websiteName: websiteName,
        storyName: storyName,
        authorName: authorName,
        chaptersCount: chaptersCount,
        lastUpdated: lastUpdated,
        summary: summary,
        storyStatus: storyStatus,
        category: category,
        genre: genre,
        rating: rating,
        storyLink: storyLink,
        authorLink: authorLink,
        totalWordCount: totalWordCount,
        chapterNames: chapterNames,
        chapterLinks: chapterLinks
    };
}

module.exports = {
    process: process
}
